avocado/core/scheduler.py

```diff
@@ -0,0 +1,143 @@
+    """Exception raised when there are no more pending tasks to be started."""
+
+
+class Scheduler:
+    """Manages the scheduling of tasks, keeping track of the their status."""
+
+    #: The interval given to other coroutines when a the scheduler needs
+    #: to wait, for instance, for the conclusion of all tasks
+    INTERVAL = 0.05
+
+    def __init__(self, tasks, parallel_tasks=None, spawner=None):
+        """Initializes a new scheduler.
+
+        :param tasks: the tasks to be scheduled to be executed
+        :type tasks: list of :class:`avocado.core.nrunner.Task`
+        :param parallel_tasks: the number of parallel tasks that will be kept
+                               running at any given time.  It defaults to
+                               twice the number of active CPUs in the local
+                               system, minus one.
+        :type parallel_tasks: int
+        :param spawner: an instance of a spawner to be used in starting tasks
+                        and checking for their "health", that is, if a task
+                        appears to be alive or not.  It defaults to a
+                       :class:`avocado.core.spawners.process.ProcessSpawner`
+        :type spawner: :class:`avocado.core.spawners.common.BaseSpawner`
+        """
+        self.pending_tasks = tasks
+
+        if parallel_tasks is None:
+            parallel_tasks = 2 * cpu_count() - 1
+        self.parallel_tasks = parallel_tasks
+
+        if spawner is None:
+            spawner = ProcessSpawner()
+        self.spawner = spawner
+        self.started_tasks = []
+        self.start_failed_tasks = []
+        #: Tasks that have finished from the manager perspective
+        self.finished_tasks = []
+
+    def __repr__(self):
+        return ('<Scheduler pending: "{}" started: "{}" start failed: {} '
+                'finished: "{}">').format(self.pending_tasks,
+                                          self.started_tasks,
+                                          self.start_failed_tasks,
+                                          self.finished_tasks)
+
+    @asyncio.coroutine
+    def tick(self):
+        """Runs one cycle of scheduler tasks, based on current parallel tasks.
+
+        This method facilitates the use of the scheduler in a loop.  It will
+        return False if new task was not spawned, which can be used to abort
+        a loop.
+
+        :rtype: bool
+        """
+        yield from self.reconcile_task_status()
+        if self.is_complete():
+            return False
+        if not self.should_spawn_new_task():
+            return False
+        try:
+            result = yield from self.spawn_next_task()
+            return result
+        except SchedulerNoPendingTasksException:
+            return False
+
+    def should_spawn_new_task(self):
+        """Whether to spawn a new task based on the current load."""
+        return len(self.started_tasks) < self.parallel_tasks
```

Comment:
What is not clear to me, right now, is how this `Scheduler` will fit on an architecture where we have a `RemoteVMSpawner` or `AWSSpawner`, for instance.

Checking for the local number of CPU is only valid for tasks that will run locally, right? Would it make sense if each `Spawner` had its own "Scheduler"? For the user perspective/use-case:  _"Ok, I would like to run 10 tasks, 2 locally, 5 on a Podman (also locally), and 3 on this AWS account."_

Avocado would be responsible to assign the tasks to the proper spawners, and each spawner would keep tracking the "available resource to run".

I can see that right now, we have only local CPUs because we have only `PodmanSpawner` and `ProcessSpawner`, but I thought that the whole idea of having spawners would be to make this as flexible as possible.

During the weekend I was thinking about how easy it would be to create an `OpenStackSpawner`, and look like this logic is making parallel the "process of start the spawners" not the "run itself". What I'm missing here?

Again, I'm still trying to understand the big picture here, so maybe this will be clear soon.

Reply:
> What is not clear to me, right now, is how this `Scheduler` will fit on an architecture where we have a `RemoteVMSpawner` or `AWSSpawner`, for instance.
> 
> Checking for the local number of CPU is only valid for tasks that will run locally, right? Would it make sense if each `Spawner` had its own "Scheduler"? For the user perspective/use-case: _"Ok, I would like to run 10 tasks, 2 locally, 5 on a Podman (also locally), and 3 on this AWS account."_
> 

We can extend the spawners to signal to the schedulers that they were not able, *yet*, to accept new tasks. For instance, raising something like `SpawnerPleaseTryAgain` or SpawnerResourceLimitReached` and the scheduler would take notice of that and attempt to reschedule it at a later time.  By concentrating that in one scheduler, it can even attempt to re-run tasks locally, instead of their originally intended remote cloud, because more local resources were  made available since then. Or vice-versa.

> Avocado would be responsible to assign the tasks to the proper spawners, and each spawner would keep tracking the "available resource to run".
> 

Yes, that's possible.

> I can see that right now, we have only local CPUs because we have only `PodmanSpawner` and `ProcessSpawner`, but I thought that the whole idea of having spawners would be to make this as flexible as possible.
> 

And it is.  I don't get how this implementation prevents such flexibility.

> During the weekend I was thinking about how easy it would be to create an `OpenStackSpawner`, and look like this logic is making parallel the "process of start the spawners" not the "run itself". What I'm missing here?
> 

You mean spawning multiple tasks in such an `OpenStackSpawner`, right? How is that not possible with the current architecture? 

> Again, I'm still trying to understand the big picture here, so maybe this will be clear soon.

Sure, I'm working as we speak on updated docs, so I hope that it will make the overall architecture proposal here clearer.

Reply:
> > What is not clear to me, right now, is how this `Scheduler` will fit on an architecture where we have a `RemoteVMSpawner` or `AWSSpawner`, for instance.
> > Checking for the local number of CPU is only valid for tasks that will run locally, right? Would it make sense if each `Spawner` had its own "Scheduler"? For the user perspective/use-case: _"Ok, I would like to run 10 tasks, 2 locally, 5 on a Podman (also locally), and 3 on this AWS account."_
> 
> We can extend the spawners to signal to the schedulers that they were not able, _yet_, to accept new tasks. For instance, raising something like `SpawnerPleaseTryAgain` or SpawnerResourceLimitReached` and the scheduler would take notice of that and attempt to reschedule it at a later time. By concentrating that in one scheduler, it can even attempt to re-run tasks locally, instead of their originally intended remote cloud, because more local resources were made available since then. Or vice-versa.

I see the idea, but I'm wondering how this would impact on the following scenario:

We have a pool with 100 virtual machines on a cloud to run tests, and the current logic says that my machine has only 8 CPUs available. This will not be a bottleneck? Because you are using the local CPU available instead of the remote machines available? 

> > I can see that right now, we have only local CPUs because we have only `PodmanSpawner` and `ProcessSpawner`, but I thought that the whole idea of having spawners would be to make this as flexible as possible.
> 
> And it is. I don't get how this implementation prevents such flexibility.

See my comment here above.
 
> > During the weekend I was thinking about how easy it would be to create an `OpenStackSpawner`, and look like this logic is making parallel the "process of start the spawners" not the "run itself". What I'm missing here?
> 
> You mean spawning multiple tasks in such an `OpenStackSpawner`, right? How is that not possible with the current architecture?

Sorry if it was not clear on the previous comment. My concern is creating a bottleneck on the local side even when we have free remote resources available.

> > Again, I'm still trying to understand the big picture here, so maybe this will be clear soon.
> 
> Sure, I'm working as we speak on updated docs, so I hope that it will make the overall architecture proposal here clearer.

That's is good, thanks.

Reply:
> > > What is not clear to me, right now, is how this `Scheduler` will fit on an architecture where we have a `RemoteVMSpawner` or `AWSSpawner`, for instance.
> > > Checking for the local number of CPU is only valid for tasks that will run locally, right? Would it make sense if each `Spawner` had its own "Scheduler"? For the user perspective/use-case: _"Ok, I would like to run 10 tasks, 2 locally, 5 on a Podman (also locally), and 3 on this AWS account."_
> > 
> > 
> > We can extend the spawners to signal to the schedulers that they were not able, _yet_, to accept new tasks. For instance, raising something like `SpawnerPleaseTryAgain` or SpawnerResourceLimitReached` and the scheduler would take notice of that and attempt to reschedule it at a later time. By concentrating that in one scheduler, it can even attempt to re-run tasks locally, instead of their originally intended remote cloud, because more local resources were made available since then. Or vice-versa.
> 
> I see the idea, but I'm wondering how this would impact on the following scenario:
> 
> We have a pool with 100 virtual machines on a cloud to run tests, and the current logic says that my machine has only 8 CPUs available. This will not be a bottleneck? Because you are using the local CPU available instead of the remote machines available, right?
> 
> Instead of trying to push tasks constantly to the spawner, why not send all and let it handle the scheduler?
> 

Yes, the current implementation only takes *one* spawner, and the number of parallel tasks is static.  We can surely attempt to cover the future needs now (multiple spawners and a dynamic number of parallel tasks), but IMO that tends to introduce over engineered solutions and goes against some agile best practices.  I very much prefer to take one small improvement at a time, unless it's clear that it prevents future needs.

Still on the scheduler .vs. spawner topic, we still need a central point to keep track of all pending tasks, which is the main role of the scheduler here.  That seems to be what you're missing here.

Reply:
Also, for me, it is a little bit strange to change the "spawner" during the runtime, since that is a request of the user:

```
$ avocado run --podman-spawner tests/*
```

The user is **explicitly** asking to run with a specific spawner.

If we attempt to run locally, might be strange. I'm wondering how this could possibly add unnecessary complexity. IMO, returning an error saying that it was not possible should be enough.

Sorry if I'm not seeing the advantages of your approach instead of proposed here. Maybe will help with the documentation, or not! ;)

Reply:
> Also, for me, it is a little bit strange to change the "spawner" during the runtime, since that is a users' request:
> 
> ```
> $ avocado run --podman-spawner tests/*
> ```
> 
> The user is **explicitly** asking to run with a specific spawner.
> 
> If we attempt to run locally, might be strange. I'm wondering how this could possibly add unnecessary complexity. IMO, returning an error saying that it was not possible should be enough.
>

I did not mention we should switch to a different spawner than the user requested.  I was following your idea of being able to cope with dynamic capacity, and evolved the idea (future possibilities) into a dynamic scheduler that could, if the user wants to, use either local or remote resources based on availability during run time. 
 
> Sorry if I'm not seeing the advantages of your approach instead of proposed here. Maybe will help with the documentation, or not! ;)

Reply:
> Yes, the current implementation only takes _one_ spawner, and the number of parallel tasks is static. We can surely attempt to cover the future needs now (multiple spawners and a dynamic number of parallel tasks), but IMO that tends to introduce over engineered solutions and goes against some agile best practices. I very much prefer to take one small improvement at a time, unless it's clear that it prevents future needs.

I see and partially agree with this approach. Maybe we could try to find a "middle ground" and still remaining agile.

This is a crucial part of the next runner and once it is being used by others might be not so easy to change. Also, it is harder to understand the overall picture/goal with individual PRs without a blueprint. I'm just trying to stress and discuss a little bit the overall idea, that is in your mind.

For sure, we don't need to implement everything now, I'm not asking this. But having a clear goal and near-future expectations, it is helpful to avoid future non-expected patches.

> Still on the scheduler .vs. spawner topic, we still need a central point to keep track of all pending tasks, which is the main role of the scheduler here. That seems to be what you're missing here.

So are we talking about a "multi-schedule architecture?"

Again, sorry if I'm stressing too much this topic, but "scheduler" seems to be a very well discussed topic outside Avocado and we could avoid implement some parts, or, at least, discuss more the overall idea a priori.

Reply:
> Yes, the current implementation only takes _one_ spawner, and the number of parallel tasks is static. We can surely attempt to cover the future needs now (multiple spawners and a dynamic number of parallel tasks), but IMO that tends to introduce over engineered solutions and goes against some agile best practices. I very much prefer to take one small improvement at a time, unless it's clear that it prevents future needs.
> 

While it is good to have a feature available to users as soon as possible, I personally do not like the idea of delivering a key component without having an overall discussion of what the architecture will look like. From personal experience, growth without a defined architecture may lock the project for new features sooner than we think.

As you mentioned the agile best practices, it is always good to remember that agile is not the synonym of doing things **fast**. It is always good to remember the Agile Manifesto: https://agilemanifesto.org/iso/en/principles.html

Reply:
> > Yes, the current implementation only takes _one_ spawner, and the number of parallel tasks is static. We can surely attempt to cover the future needs now (multiple spawners and a dynamic number of parallel tasks), but IMO that tends to introduce over engineered solutions and goes against some agile best practices. I very much prefer to take one small improvement at a time, unless it's clear that it prevents future needs.
> 
> While it is good to have a feature available to users as soon as possible, I personally do not like the idea of delivering a key component without having an overall discussion of what the architecture will look like. From personal experience, growth without a defined architecture may lock the project for new features sooner than we think.
> 
> As you mentioned the agile best practices, it is always good to remember that agile is not the synonym of doing things **fast**. It is always good to remember the Agile Manifesto: https://agilemanifesto.org/iso/en/principles.html

I hear you, and I agree that a better description of the architecture will help us all.  I do accept the blame for not doing it before, and FIY I'm actively working on it.

On the other hand, the let me try to make one thing clear about my core beliefs: I care about fulfilling the user/customer requirements 100x more than I care about implementation details, or the needs to refactor things later. Value is what matters, and this happens to be, IMO the core of the agile principles.

So, when look at any code proposal, I always think at how it will enable a user/customer requirement, and thus real value.  I'm mentioning this because I think the bullet points under the motivations for the N(ext)Runner are the most important things to consider.

About the scheduler, talking about "multiple schedulers" or "spawners should be schedulers", affects (but not prevents) only one of the bullet points there ("Distributed execution of tests across a pool of any combination of  processes, containers, virtual machines, etc."), while it contributes to enable and deliver many of the bullet points. 

Bear in mind that I'm not advocating for sketchy and fragile code, quite the contrary. This PR, for instance adds a good number of tests. I do think Avocado's code is reasonably good and is getting better, with more lint checks getting enabled, almost 1000 selftests (besides the release job, deployment checks, etc).

And finally, I do think we need to move fast, but never rush nor panic (too much of that in the world, right?).

Reply:
BTW, as a simple sign of good faith:

https://github.com/avocado-framework/avocado/pull/3773

Expect more soon.

Reply:
@clebergnu I think that any improvement in the current documentation is a good thing. Thanks for that.

But I was expecting a more "detailed" discussion on the "Spawners vs Schedulers" architecture/design.

Since that the initial draft of the Next Runner architecture didn't mention neither Spawners nor Schedulers, it looks like we need a blueprint just to improve the "Spawners vs Schedulers" discussion. 

When issue #3754 was open I was hoping for a blueprint to start a new discussion on it.

Reply:
> @clebergnu I think that any improvement in the current documentation is a good thing. Thanks for that.
> 
> But I was expecting a more "detailed" discussion on the "Spawners vs Schedulers" architecture/design.
> 
> Since that the initial draft of the Next Runner architecture didn't mention neither Spawners nor Schedulers, it looks like we need a blueprint just to improve the "Spawners vs Schedulers" discussion.
> 
> When issue #3754 was open I was hoping for a blueprint to start a new discussion on it.

Sure, I'm filling in the voids in the current docs, and then we'll hopefully be aligned to discuss the next features.  That's why I said "expect more soon".

Reply:
> Sure, I'm filling in the voids in the current docs, and then we'll hopefully be aligned to discuss the next features. That's why I said "expect more soon".

Cool, thanks for that! While you fill in the gaps, I'm taking the chance to refresh my memory on schedulers with the classic "Scheduling, Theory, Algorithms, and Systems" from Michael Pinedo.